dxp/react_rendering.md

```diff
@@ -0,0 +1,148 @@
+# React Rendering in DXP
+
+This article describes DXP infrastructure to render React components from the server side perspective. This includes taglibs as well as low-level services.
+
+## React Renderer Service
+
+-   OSGi bundle: `com.liferay.portal.template.react.renderer.api`
+-   Interface name: `com.liferay.portal.template.react.renderer.ReactRenderer`
+-   Service type: `singleton`
+
+This is the low-level service to render React components.
+
+> Note that it renders React components, not any React code, so you must provide a source which exports a React component to render.
+
+It has one method:
+
+```java
+public void renderReact(
+		ComponentDescriptor componentDescriptor, Map<String, Object> data,
+		HttpServletRequest httpServletRequest, Writer writer)
+	throws IOException;
+```
+
+which receives:
+
+1. A `ComponentDescriptor` describing the source of the React component to render. More on this in the next section.
+2. A `Map` containing arbitrary data that will be injected to the React component as initial properties. Note that the data inside the map must be compatible with JavaScript, so you cannot inject any Java object you like, only primitive types, `Map`s, `Array`s, ... that can be serialized to JSON.
+3. The current `HttpServletRequest` for which the component is being rendered.
+4. A `Writer` where the rendered HTML (and optionally the JavaScript) is written to.
+
+Note that parameter 4 is not intended to be used to write the rendered HTML and JavaScript to any place you like but to pass the `PrintWriter` from the `HttpServletResponse` or the `JspWriter` from the `PageContext` you have available when invoking the method.
+
+Also, take into account that, if you don't render the JavaScript inline (see option `positionInline` of `ComponentDescriptor`), the JavaScript will be written to the bottom of the current `HttpServletResponse`, while the HTML will be inlined into the passed `Writer`.
+
+### What the ComponentDescriptor contains
+
+`ComponentDescriptor` instances are [value objects](https://en.wikipedia.org/wiki/Value_object) that describe what needs to be rendered. A `ComponentDescriptor` may have up to five fields (of which only the first one is mandatory):
+
+1. `module`: the name of the AMD module that exports the React component to be rendered. Note that you need to provide the canonical name of the module (including its version number). To avoid hardcoding any version number you may use the instance of the `com.liferay.frontend.js.loader.modules.extender.npm.NPMResolver` service tied to the OSGi bundle containing the AMD module to be rendered.
```

Comment:
Maybe add an example of this? Both in what a `module` usually looks like: `foo@1.3.2/js/Foo.js` and how to get it `NPMResolver.resolveModule('Foo')`.